localstack-core/localstack/services/apigateway/next_gen/execute_api/header_utils.py

```diff
@@ -32,3 +101,108 @@ def build_multi_value_headers(headers: Headers) -> dict[str, list[str]]:
+        if headers.get(header):
+            LOG.debug("Dropping header: %s", header)
+            headers.pop(header)
+
+
+def drop_invocation_headers(headers: Headers, integration_type: IntegrationType):
+    """Will modify the provided headers in-place. Dropping matching headers from the provided integration type"""
+    _drop_headers(headers, DROPPED_FROM_REQUEST_PROXY_COMMON)
+
+
+def drop_request_headers(headers: Headers, integration_type: IntegrationType):
+    """Will modify the provided headers in-place. Dropping matching headers for the provided integration type"""
+    match integration_type:
+        case IntegrationType.AWS:
+            _drop_headers(headers, DROPPED_FROM_INTEGRATION_REQUESTS_AWS)
+        case IntegrationType.HTTP | IntegrationType.HTTP_PROXY:
+            _drop_headers(headers, DROPPED_FROM_INTEGRATION_REQUESTS_HTTP)
+        case _:
+            _drop_headers(headers, DROPPED_FROM_INTEGRATION_REQUESTS_COMMON)
+
+
+def drop_response_headers(headers: Headers, integration_type: IntegrationType):
+    """Will modify the provided headers in-place. Dropping matching headers for the provided integration type"""
+    match integration_type:
+        case IntegrationType.HTTP_PROXY:
+            _drop_headers(headers, DROPPED_FROM_INTEGRATION_RESPONSES_HTTP_PROXY)
+        case _:
+            _drop_headers(headers, DROPPED_FROM_INTEGRATION_RESPONSES_COMMON)
+
+
+def remap_response_headers(headers: Headers, integration_type: IntegrationType):
+    """Remaps the provided headers in-place. Adding new `x-amzn-Remapped-` headers and dropping the original headers"""
+    match integration_type:
+        case IntegrationType.HTTP | IntegrationType.AWS:
+            to_remap = REMAPPED_FROM_INTEGRATION_RESPONSE_NON_PROXY
+        case _:
+            to_remap = REMAPPED_FROM_INTEGRATION_RESPONSE_COMMON
+
+    for header in to_remap:
+        if headers.get(header):
+            LOG.debug("Remapping header: %s", header)
+            remapped = headers.pop(header)
+            headers[f"x-amzn-Remapped-{header}"] = remapped
+
+
+def validate_request_headers(headers: dict[str, str], integration_type: IntegrationType):
+    """Validates and raises an error when attempting to set an illegal header"""
+    to_validate = ILLEGAL_INTEGRATION_REQUESTS_COMMON
+    match integration_type:
+        case IntegrationType.AWS | IntegrationType.AWS_PROXY:
+            to_validate = ILLEGAL_INTEGRATION_REQUESTS_AWS
+
+    for header in headers.keys():
+        if header.lower() in to_validate:
+            LOG.debug(
+                "Execution failed due to configuration error: %s header already present", header
+            )
+            raise InternalServerError("Internal server error")
+
+
+def set_default_headers(headers: Headers, default_headers: dict[str, str]):
+    for header, value in default_headers.items():
+        if not headers.get(header):
+            headers.set(header, value)
+
+
+def set_default_response_headers(
```

Comment:
this one being used for both exception and invocation feels so close to a Response Handler, I'm tempted 😄

Reply:
The problem with that is that `Content-Type: text/plain; charset=utf-8` is added automatically when creating a `Response` so it would make it tricky know if it was set or not. That is the only thing that prevented me from creating a response handler for it.
Welcome to suggestion! :wink:

Reply:
Maybe we should set at least the `Content-Type` as the default value of the `Response` object? Before calling `response.update_from`? As we know AWS will always return `application/json` if not "overriden"? so we just override the `Response`, and if the request parameter/etc overrides it then it's alright?